airbyte-lib/airbyte_lib/caches/base.py

```diff
@@ -0,0 +1,591 @@
+                )
+                self._connection_to_reuse = bootstrap_engine.connect()
+
+            self._engine = create_engine(
+                sql_alchemy_url,
+                creator=lambda: self._connection_to_reuse,
+                poolclass=StaticPool,
+                echo=DEBUG_MODE,
+                # isolation_level="AUTOCOMMIT",
+            )
+        else:
+            # Regular engine creation for new connections
+            self._engine = create_engine(
+                sql_alchemy_url,
+                echo=DEBUG_MODE,
+                # isolation_level="AUTOCOMMIT",
+            )
+
+        return self._engine
+
+    @contextmanager
+    def get_sql_connection(self) -> Generator[sqlalchemy.engine.Connection, None, None]:
+        """A context manager which returns a new SQL connection for running queries.
+
+        If the connection needs to close, it will be closed automatically.
+        """
+        if self.use_singleton_connection and self._connection_to_reuse is not None:
+            connection = self._connection_to_reuse
+            yield connection
+
+        else:
+            with self.get_sql_engine().begin() as connection:
+                yield connection
+
+        if not self.use_singleton_connection:
+            connection.close()
+            del connection
+
+    def get_sql_table_name(
+        self,
+        stream_name: str,
+    ) -> str:
+        """Return the name of the SQL table for the given stream."""
+        return self._normalize_table_name(
+            f"{self.config.table_prefix}{stream_name}{self.config.table_suffix}",
+        )
+
+    @final
+    def get_sql_table(
+        self,
+        stream_name: str,
+    ) -> sqlalchemy.Table:
+        """Return a temporary table name."""
+        table_name = self.get_sql_table_name(stream_name)
+        return sqlalchemy.Table(
+            table_name,
+            sqlalchemy.MetaData(schema=self.config.schema_name),
+            autoload_with=self.get_sql_engine(),
+        )
+
+    # Read methods:
+
+    def get_stream_records(
+        self,
+        stream_name: str,
+    ) -> Iterable[dict[str, Any]]:
+        """Uses SQLAlchemy to select all rows from the table."""
+        table_ref = self.get_sql_table(stream_name)
+        stmt = table_ref.select()
+        with self.get_sql_connection() as conn:
+            yield from conn.execute(stmt)
+
+    def get_pandas_dataframe(
+        self,
+        stream_name: str,
+    ) -> pd.DataFrame:
+        """Return a Pandas data frame with the stream's data."""
+        table_name = self.get_sql_table_name(stream_name)
+        engine = self.get_sql_engine()
+        return pd.read_sql_table(table_name, engine)
+
+    # Protected members (non-public interface):
+
+    def _ensure_schema_exists(
+        self,
+    ) -> None:
+        """Return a new (unique) temporary table name."""
+        schema_name = self.config.schema_name
+        if schema_name in self._get_schemas_list():
```

Comment:
This makes sure the table exists, but it won't make sure the column types are set up correctly. Should we add this here?

Reply:
Yes, but probably out of scope for this iteration. Adding as a TODO.

Reply:
I've added a new `_ensure_compatible_table_schema` method to handle this. For now, it just checks all columns exist and (if needed), the primary keys are defined.